src/Propulsion.SqlStreamStore/StreamReader.fs

```diff
@@ -0,0 +1,207 @@
+                    len0ToNull data,
+                    len0ToNull metadata,
+                    msg.MessageId,
+                    timestamp = DateTimeOffset(msg.CreatedUtc)
+                )
+
+            { stream = StreamName.internalParseSafe msg.StreamId; event = event }
+
+    type Stats(logger: ILogger, statsInterval: TimeSpan) =
+
+        let mutable batchFirstPosition = 0L
+        let mutable batchLastPosition = 0L
+        let mutable batchCaughtUp = false
+
+        let mutable pagesRead = 0
+        let mutable pagesEmpty = 0
+        let mutable recentPagesRead = 0
+        let mutable recentPagesEmpty = 0
+
+        let mutable currentBatches = 0
+        let mutable maxBatches = 0
+
+        let mutable lastCommittedPosition = 0L
+
+        let report () =
+            logger.Information(
+                "Pages Read {pagesRead} Empty {pagesEmpty} | Recent Read {recentPagesRead} Empty {recentPagesEmpty} | Position Read {batchLastPosition} Committed {lastCommittedPosition} | Caught up {caughtUp} | cur {cur} / max {max}",
+                pagesRead, pagesEmpty, recentPagesRead, recentPagesEmpty, batchLastPosition, lastCommittedPosition, batchCaughtUp, currentBatches, maxBatches)
+
+        member this.UpdateBatch (batch: InternalBatch) =
+            batchFirstPosition <- batch.firstPosition
+            batchLastPosition <- batch.lastPosition
+            batchCaughtUp <- batch.isEnd
+
+            pagesRead <- pagesRead + 1
+            recentPagesRead <- recentPagesRead + 1
+
+        member this.UpdateEmptyPage () =
+            pagesEmpty <- pagesEmpty + 1
+            recentPagesEmpty <- recentPagesEmpty + 1
+
+        member this.UpdateCommitedPosition(pos) =
+            lastCommittedPosition <- pos
+
+        member this.UpdateCurMax(cur, max) =
+            currentBatches <- cur
+            maxBatches <- max
+
+        member this.Start =
+            async {
+                let! ct = Async.CancellationToken
+                while not ct.IsCancellationRequested do
+                    report ()
+                    recentPagesRead <- 0
+                    recentPagesEmpty <- 0
+                    do! Async.Sleep statsInterval
+            }
+
+    [<RequireQualifiedAccess>]
+    [<NoComparison>]
+    type Work =
+        | TakeInitial
+        | Page of ReadAllPage
+        | TakeNext of ReadAllPage
+
+    type SubmitBatchHandler =
+        // ingester submit method: epoch * checkpoint * items -> write result
+        int64 * Async<unit> * seq<Propulsion.Streams.StreamEvent<byte[]>> -> Async<int*int>
+
+type StreamReader
+    (
+        logger: ILogger,
+        store: IStreamStore,
+        checkpointer: ICheckpointer,
+        submitBatch: SubmitBatchHandler,
+        streamId,
+        consumerGroup,
+        maxBatchSize: int,
+        tailSleepInterval: TimeSpan,
+        statsInterval: TimeSpan
+    ) =
+
+    let stats = Stats(logger, statsInterval)
+
+    let commit batch =
```

Comment:
only lastPosition is used from `batch` ?

Reply:
As in, just pass position here? Can be done.

Reply:
👍